README.md
# vit-keras
This is a Keras implementation of the models described in [An Image is Worth 16x16 Words:
Transformes For Image Recognition at Scale](https://arxiv.org/pdf/2010.11929.pdf). It is based on an earlier implementation from [tuvovan](https://github.com/tuvovan/Vision_Transformer_Keras), modified to match the Flax implementation in the [official repository](https://github.com/google-research/vision_transformer).

The weights here are ported over from the weights provided in the official repository. See `utils.load_weights_numpy` to see how this is done (it's not pretty, but it does the job).

## Usage
Install this package using `pip install vit-keras`

You can use the model out-of-the-box with ImageNet 2012 classes using
something like the following. The weights will be downloaded automatically.

```python
from vit_keras import vit, utils

image_size = 384
classes = utils.get_imagenet_classes()
model = vit.vit_b16(
    image_size=image_size,
    activation='sigmoid',
    pretrained=True,
    include_top=True,
    pretrained_top=True
)
url = 'https://upload.wikimedia.org/wikipedia/commons/d/d7/Granny_smith_and_cross_section.jpg'
image = utils.read(url, image_size)
X = vit.preprocess_inputs(image).reshape(1, image_size, image_size, 3)
y = model.predict(X)
print(classes[y[0].argmax()]) # Granny smith
```

You can fine-tune using a model loaded as follows.

```python
image_size = 224
model = vit.vit_l32(
    image_size=image_size,
    activation='sigmoid',
    pretrained=True,
    include_top=True,
    pretrained_top=False,
    classes=200
)
# Train this model on your data as desired.
```

## Visualizing Attention Maps
There's some functionality for plotting attention maps for a given image and model. See example below. I'm not sure I'm doing this correctly (the official repository didn't have example code). Feedback /corrections welcome!

```python
import numpy as np
import matplotlib.pyplot as plt
from vit_keras import vit, utils, visualize

# Load a model
image_size = 384
classes = utils.get_imagenet_classes()
model = vit.vit_b16(
    image_size=image_size,
    activation='sigmoid',
    pretrained=True,
    include_top=True,
    pretrained_top=True
)
classes = utils.get_imagenet_classes()

# Get an image and compute the attention map
url = 'https://upload.wikimedia.org/wikipedia/commons/b/bc/Free%21_%283987584939%29.jpg'
image = utils.read(url, image_size)
attention_map = visualize.attention_map(model=model, image=image)
print('Prediction:', classes[
    model.predict(vit.preprocess_inputs(image)[np.newaxis])[0].argmax()]
)  # Prediction: Eskimo dog, husky

# Plot results
fig, (ax1, ax2) = plt.subplots(ncols=2)
ax1.axis('off')
ax2.axis('off')
ax1.set_title('Original')
ax2.set_title('Attention Map')
_ = ax1.imshow(image)
_ = ax2.imshow(attention_map)
```

![example of attention map](https://raw.githubusercontent.com/faustomorales/vit-keras/master/docs/attention_map_example.jpg)


CHANGELOG.md
# v0.0.1

- Initial version of project.


vit_keras/utils.py
import os
import typing
import warnings
from urllib import request
from http import client
import io
import pkg_resources
import validators
import numpy as np
import scipy as sp
import cv2

try:
    import PIL
    import PIL.Image
except ImportError:  # pragma: no cover
    PIL = None

ImageInputType = typing.Union[str, np.ndarray, "PIL.Image.Image", io.BytesIO]


def get_imagenet_classes() -> typing.List[str]:
    """Get the list of ImageNet 2012 classes."""
    filepath = pkg_resources.resource_filename("vit_keras", "imagenet2012.txt")
    with open(filepath) as f:
        classes = [l.strip() for l in f.readlines()]
    return classes


def read(filepath_or_buffer: ImageInputType, size, timeout=None):
    """Read a file into an image object
    Args:
        filepath_or_buffer: The path to the file or any object
            with a `read` method (such as `io.BytesIO`)
        size: The size to resize the image to.
        timeout: If filepath_or_buffer is a URL, the timeout to
            use for making the HTTP request.
    """
    if PIL is not None and isinstance(filepath_or_buffer, PIL.Image.Image):
        return np.array(filepath_or_buffer.convert("RGB"))
    if isinstance(filepath_or_buffer, (io.BytesIO, client.HTTPResponse)):
        image = np.asarray(bytearray(filepath_or_buffer.read()), dtype=np.uint8)
        image = cv2.imdecode(image, cv2.IMREAD_UNCHANGED)
    elif isinstance(filepath_or_buffer, str) and validators.url(filepath_or_buffer):
        return read(request.urlopen(filepath_or_buffer, timeout=timeout), size=size)
    else:
        if not os.path.isfile(filepath_or_buffer):
            raise FileNotFoundError(
                "Could not find image at path: " + filepath_or_buffer
            )
        image = cv2.imread(filepath_or_buffer)
    if image is None:
        raise ValueError(f"An error occurred reading {filepath_or_buffer}.")
    # We use cvtColor here instead of just ret[..., ::-1]
    # in order to ensure that we provide a contiguous
    # array for later processing. Some hashers use ctypes
    # to pass the array and non-contiguous arrays can lead
    # to erroneous results.
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    return cv2.resize(image, (size, size))


def load_weights_numpy(model, params_path, pretrained_top):
    """Load weights saved using Flax as a numpy array.

    Args:
        model: A Keras model to load the weights into.
        params_path: Filepath to a numpy archive.
        pretrained_top: Whether to load the top layer weights.
    """
    params_dict = np.load(params_path, allow_pickle=False)
    weights_dict = {}
    input_keys = list(params_dict.keys())
    pre_logits = any(l.name == "pre_logits" for l in model.layers)
    input_keys_taken = []
    output_keys_assigned = []
    output_weights = {w.name: w.numpy() for w in model.weights}
    if not pretrained_top:
        for key in ["head/kernel", "head/bias"]:
            if key in output_weights:
                del output_weights[f"{key}:0"]
            if key in input_keys:
                input_keys.remove(key)
    n_transformers = len(
        set(
            "/".join(k.split("/")[:2])
            for k in input_keys
            if k.startswith("Transformer/encoderblock_")
        )
    )
    n_transformers_out = len(
        set(
            "/".join(k.split("/")[:2])
            for k in output_weights
            if k.startswith("Transformer/encoderblock_")
        )
    )
    assert n_transformers == n_transformers_out, (
        f"Wrong number of transformers ("
        f"{n_transformers_out} in model vs. {n_transformers} in weights)."
    )

    def apply_weights(keyi, keyo):
        assert keyo in output_weights, f"{keyo} not in output weights."
        expected_shape = output_weights[keyo].shape
        actual_weights = params_dict[keyi]
        if keyi == "Transformer/posembed_input/pos_embedding":
            if expected_shape != actual_weights.shape:
                token, grid = actual_weights[0, :1], actual_weights[0, 1:]
                sin = int(np.sqrt(grid.shape[0]))
                sout = int(np.sqrt(expected_shape[1] - 1))
                warnings.warn(
                    "Resizing position embeddings from " f"{sin} to {sout}",
                    UserWarning,
                )
                zoom = (sout / sin, sout / sin, 1)
                grid = sp.ndimage.zoom(
                    grid.reshape(sin, sin, -1), zoom, order=1
                ).reshape(sout * sout, -1)
                actual_weights = np.concatenate([token, grid], axis=0)[np.newaxis]
        if "MultiHeadDotProductAttention" in keyi:
            actual_weights = actual_weights.reshape(expected_shape)
        actual_shape = actual_weights.shape
        assert expected_shape == actual_shape, (
            f"Shapes for layer: {keyi} / {keyo} do not match "
            f"({actual_shape} in weights vs. {expected_shape} in model."
        )
        weights_dict[keyo] = actual_weights
        input_keys_taken.append(keyi)
        output_keys_assigned.append(keyo)

    for tidx in range(n_transformers):
        for norm in ["LayerNorm_0", "LayerNorm_2"]:
            for inname, outname in [("scale", "gamma"), ("bias", "beta")]:
                apply_weights(
                    f"Transformer/encoderblock_{tidx}/{norm}/{inname}",
                    f"Transformer/encoderblock_{tidx}/{norm}/{outname}:0",
                )
        for mlpdense in [0, 1]:
            for subname in ["kernel", "bias"]:
                apply_weights(
                    f"Transformer/encoderblock_{tidx}/MlpBlock_3/Dense_{mlpdense}/{subname}",
                    f"Transformer/encoderblock_{tidx}/MlpBlock_3/Dense_{mlpdense}/{subname}:0",
                )
        for attvar in ["query", "key", "value", "out"]:
            for subname in ["kernel", "bias"]:
                apply_weights(
                    f"Transformer/encoderblock_{tidx}/MultiHeadDotProductAttention_1/{attvar}/{subname}",
                    f"Transformer/encoderblock_{tidx}/MultiHeadDotProductAttention_1/{attvar}/{subname}:0",
                )
    for attvar in ["embedding", "head", "pre_logits"]:
        if attvar == "head" and not pretrained_top:
            continue
        if attvar == "pre_logits" and not pre_logits:
            continue
        for subname in ["kernel", "bias"]:
            apply_weights(f"{attvar}/{subname}", f"{attvar}/{subname}:0")
    for attname in ["Transformer/posembed_input/pos_embedding"]:
        apply_weights(f"{attname}", f"{attname}:0")
    for iname, outname in [("cls", "class_token/cls:0")]:
        apply_weights(iname, outname)
    for inname, outname in [("scale", "gamma"), ("bias", "beta")]:
        apply_weights(
            f"Transformer/encoder_norm/{inname}",
            f"Transformer/encoder_norm/{outname}:0",
        )
    unused = [k for k in input_keys if k not in input_keys_taken]
    missing = [k for k in output_weights if k not in output_keys_assigned]
    if unused:
        warnings.warn(f"Did not use the following weights: {unused}", UserWarning)
    if missing:
        warnings.warn(f"Did not use the following weights: {missing}", UserWarning)
    weights = [weights_dict.get(k, output_weights[k]) for k in output_weights]
    model.set_weights(weights)


vit_keras/layers.py
# pylint: disable=arguments-differ,missing-function-docstring,missing-class-docstring

import tensorflow as tf
import tensorflow_addons as tfa


class ClassToken(tf.keras.layers.Layer):
    """Append a class token to an input layer."""

    def build(self, input_shape):
        cls_init = tf.zeros_initializer()
        self.cls = tf.Variable(
            name="cls",
            initial_value=cls_init(shape=(1, 1, input_shape[-1]), dtype="float32"),
            trainable=True,
        )

    def call(self, inputs):
        return tf.concat([self.cls, inputs], 1)


class AddPositionEmbs(tf.keras.layers.Layer):
    """Adds (optionally learned) positional embeddings to the inputs."""

    def build(self, input_shape):
        assert (
            len(input_shape) == 3
        ), f"Number of dimensions should be 3, got {len(input_shape)}"
        self.pe = tf.Variable(
            name="pos_embedding",
            initial_value=tf.initializers.random_normal(stddev=0.06)(
                shape=(1, input_shape[1], input_shape[2])
            ),
            dtype="float32",
            trainable=True,
        )

    def call(self, inputs):
        return inputs + self.pe


class MultiHeadSelfAttention(tf.keras.layers.Layer):
    def __init__(self, *args, num_heads, **kwargs):
        super().__init__(*args, **kwargs)
        self.num_heads = num_heads

    def build(self, input_shape):
        hidden_size = input_shape[-1]
        num_heads = self.num_heads
        if hidden_size % num_heads != 0:
            raise ValueError(
                f"embedding dimension = {hidden_size} should be divisible by number of heads = {num_heads}"
            )
        self.projection_dim = hidden_size // num_heads
        self.query_dense = tf.keras.layers.Dense(hidden_size, name="query")
        self.key_dense = tf.keras.layers.Dense(hidden_size, name="key")
        self.value_dense = tf.keras.layers.Dense(hidden_size, name="value")
        self.combine_heads = tf.keras.layers.Dense(hidden_size, name="out")

    # pylint: disable=no-self-use
    def attention(self, query, key, value):
        score = tf.matmul(query, key, transpose_b=True)
        dim_key = tf.cast(tf.shape(key)[-1], tf.float32)
        scaled_score = score / tf.math.sqrt(dim_key)
        weights = tf.nn.softmax(scaled_score, axis=-1)
        output = tf.matmul(weights, value)
        return output, weights

    def separate_heads(self, x, batch_size):
        x = tf.reshape(x, (batch_size, -1, self.num_heads, self.projection_dim))
        return tf.transpose(x, perm=[0, 2, 1, 3])

    def call(self, inputs):
        batch_size = tf.shape(inputs)[0]
        hidden_size = tf.shape(inputs)[-1]
        query = self.query_dense(inputs)
        key = self.key_dense(inputs)
        value = self.value_dense(inputs)
        query = self.separate_heads(query, batch_size)
        key = self.separate_heads(key, batch_size)
        value = self.separate_heads(value, batch_size)

        attention, weights = self.attention(query, key, value)
        attention = tf.transpose(attention, perm=[0, 2, 1, 3])
        concat_attention = tf.reshape(attention, (batch_size, -1, hidden_size))
        output = self.combine_heads(concat_attention)
        return output, weights


class TransformerBlock(tf.keras.layers.Layer):
    """Implements a Transformer block."""

    def __init__(self, *args, num_heads, mlp_dim, dropout, **kwargs):
        super().__init__(*args, **kwargs)
        self.num_heads = num_heads
        self.mlp_dim = mlp_dim
        self.dropout = dropout

    def build(self, input_shape):
        self.att = MultiHeadSelfAttention(
            num_heads=self.num_heads,
            name="MultiHeadDotProductAttention_1",
        )
        self.mlpblock = tf.keras.Sequential(
            [
                tf.keras.layers.Dense(
                    self.mlp_dim, activation=tfa.activations.gelu, name="Dense_0"
                ),
                tf.keras.layers.Dropout(self.dropout),
                tf.keras.layers.Dense(input_shape[-1], name="Dense_1"),
                tf.keras.layers.Dropout(self.dropout),
            ],
            name="MlpBlock_3",
        )
        self.layernorm1 = tf.keras.layers.LayerNormalization(
            epsilon=1e-6, name="LayerNorm_0"
        )
        self.layernorm2 = tf.keras.layers.LayerNormalization(
            epsilon=1e-6, name="LayerNorm_2"
        )
        self.dropout = tf.keras.layers.Dropout(self.dropout)

    def call(self, inputs, training):
        x = self.layernorm1(inputs)
        x, weights = self.att(x)
        x = self.dropout(x, training=training)
        x = x + inputs
        y = self.layernorm2(x)
        y = self.mlpblock(y)
        return x + y, weights


vit_keras/visualize.py
import cv2
import numpy as np
import tensorflow as tf
from . import layers, vit


def attention_map(model, image):
    """Get an attention map for an image and model using the technique
    described in Appendix D.7 in the paper (unofficial).

    Args:
        model: A ViT model
        image: An image for which we will compute the attention map.
    """
    size = model.input_shape[1]
    grid_size = int(np.sqrt(model.layers[5].output_shape[0][-2] - 1))

    # Prepare the input
    X = vit.preprocess_inputs(cv2.resize(image, (size, size)))[np.newaxis, :]  # type: ignore

    # Get the attention weights from each transformer.
    outputs = [
        l.output[1] for l in model.layers if isinstance(l, layers.TransformerBlock)
    ]
    weights = np.array(
        tf.keras.models.Model(inputs=model.inputs, outputs=outputs).predict(X)
    )
    num_layers = weights.shape[0]
    num_heads = weights.shape[2]
    reshaped = weights.reshape(
        (num_layers, num_heads, grid_size ** 2 + 1, grid_size ** 2 + 1)
    )

    # From Appendix D.6 in the paper ...
    # Average the attention weights across all heads.
    reshaped = reshaped.mean(axis=1)

    # From Section 3 in https://arxiv.org/pdf/2005.00928.pdf ...
    # To account for residual connections, we add an identity matrix to the
    # attention matrix and re-normalize the weights.
    reshaped = reshaped + np.eye(reshaped.shape[1])
    reshaped = reshaped / reshaped.sum(axis=(1, 2))[:, np.newaxis, np.newaxis]

    # Recursively multiply the weight matrices
    v = reshaped[-1]
    for n in range(1, len(reshaped)):
        v = np.matmul(v, reshaped[-1 - n])

    # Attention from the output token to the input space.
    mask = v[0, 1:].reshape(grid_size, grid_size)
    mask = cv2.resize(mask / mask.max(), image.shape[:2])[..., np.newaxis]
    return (mask * image).astype("uint8")


vit_keras/vit.py
# type: ignore
import tensorflow as tf
from . import layers, utils

CONFIG_B = {
    "dropout": 0.1,
    "mlp_dim": 3072,
    "num_heads": 12,
    "num_layers": 12,
    "hidden_size": 768,
}

CONFIG_L = {
    "dropout": 0.1,
    "mlp_dim": 4096,
    "num_heads": 16,
    "num_layers": 24,
    "hidden_size": 1024,
}

BASE_URL = "https://github.com/faustomorales/vit-keras/releases/download/dl/"

WEIGHTS = {
    False: {
        "B16": (
            BASE_URL + "ViT-B_16_imagenet21k.npz",
            "ViT-B_16_imagenet21k.npz",
        ),
        "B32": (BASE_URL + "ViT-B_32_imagenet21k.npz", "ViT-B_32_imagenet21k.npz"),
        # We're using the fine-tuned weights here because the non-fine-tuned weights
        # are not available yet. See https://github.com/googlse-research/vision_transformer/issues/15
        "L16": (
            BASE_URL + "ViT-L_16_imagenet21k+imagenet2012.npz",
            "ViT-L_16_imagenet21k+imagenet2012.npz",
        ),
        "L32": (BASE_URL + "ViT-L_32_imagenet21k.npz", "ViT-L_32_imagenet21k.npz"),
    },
    True: {
        "B16": (
            BASE_URL + "ViT-B_16_imagenet21k+imagenet2012.npz",
            "ViT-B_16_imagenet21k+imagenet2012.npz",
        ),
        "B32": (
            BASE_URL + "ViT-B_32_imagenet21k+imagenet2012.npz",
            "ViT-B_32_imagenet21k+imagenet2012.npz",
        ),
        "L16": (
            BASE_URL + "ViT-L_16_imagenet21k+imagenet2012.npz",
            "ViT-L_16_imagenet21k+imagenet2012.npz",
        ),
        "L32": (
            BASE_URL + "ViT-L_32_imagenet21k+imagenet2012.npz",
            "ViT-L_32_imagenet21k+imagenet2012.npz",
        ),
    },
}


def preprocess_inputs(X):
    """Preprocess images"""
    return tf.keras.applications.imagenet_utils.preprocess_input(
        X, data_format=None, mode="tf"
    )


def build_model(
    image_size: int,
    patch_size: int,
    num_layers: int,
    hidden_size: int,
    num_heads: int,
    mlp_dim: int,
    classes: int,
    dropout=0.1,
    activation="linear",
    include_top=True,
    representation_size=None,
):
    """Build a ViT model.

    Args:
        image_size: The size of input images.
        patch_size: The size of each patch (must fit evenly in image_size)
        classes: optional number of classes to classify images
            into, only to be specified if `include_top` is True, and
            if no `weights` argument is specified.
        num_layers: The number of transformer layers to use.
        hidden_size: The number of filters to use
        num_heads: The number of transformer heads
        mlp_dim: The number of dimensions for the MLP output in the transformers.
        dropout_rate: fraction of the units to drop for dense layers.
        activation: The activation to use for the final layer.
        include_top: Whether to include the final classification layer. If not,
            the output will have dimensions (batch_size, hidden_size).
        representation_size: The size of the representation prior to the
            classification layer. If None, no Dense layer is inserted.
    """
    assert image_size % patch_size == 0, "image_size must be a multiple of patch_size"
    x = tf.keras.layers.Input(shape=(image_size, image_size, 3))
    y = tf.keras.layers.Conv2D(
        filters=hidden_size,
        kernel_size=patch_size,
        strides=patch_size,
        padding="valid",
        name="embedding",
    )(x)
    y = tf.keras.layers.Reshape((-1, hidden_size))(y)
    y = layers.ClassToken(name="class_token")(y)
    y = layers.AddPositionEmbs(name="Transformer/posembed_input")(y)
    for n in range(num_layers):
        y, _ = layers.TransformerBlock(
            num_heads=num_heads,
            mlp_dim=mlp_dim,
            dropout=dropout,
            name=f"Transformer/encoderblock_{n}",
        )(y)
    y = tf.keras.layers.LayerNormalization(
        epsilon=1e-6, name="Transformer/encoder_norm"
    )(y)
    y = tf.keras.layers.Lambda(lambda v: v[:, 0], name="ExtractToken")(y)
    if representation_size is not None:
        y = tf.keras.layers.Dense(
            representation_size, name="pre_logits", activation="tanh"
        )(y)
    if include_top:
        y = tf.keras.layers.Dense(classes, name="head", activation=activation)(y)
    return tf.keras.models.Model(inputs=x, outputs=y)


def load_pretrained(key, model, pretrained_top):
    """Load model weights for a known configuration."""
    origin, fname = WEIGHTS[pretrained_top][key]
    local_filepath = tf.keras.utils.get_file(fname, origin, cache_subdir="weights")
    utils.load_weights_numpy(model, local_filepath, pretrained_top)


def vit_b16(
    image_size: int = 224,
    classes=1000,
    activation="linear",
    include_top=True,
    pretrained=True,
    pretrained_top=True,
):
    """Build ViT-B16. All arguments passed to build_model."""
    if pretrained_top:
        assert classes == 1000, "Can only use pretrained_top if classes = 1000."
        assert include_top, "Can only use pretrained_top with include_top."
        assert pretrained, "Can only use pretrained_top with pretrained."
    model = build_model(
        **CONFIG_B,
        patch_size=16,
        image_size=image_size,
        classes=classes,
        activation=activation,
        include_top=include_top,
        representation_size=768 if pretrained and not pretrained_top else None,
    )
    if pretrained:
        load_pretrained(key="B16", model=model, pretrained_top=pretrained_top)
    return model


def vit_b32(
    image_size: int = 224,
    classes=1000,
    activation="linear",
    include_top=True,
    pretrained=True,
    pretrained_top=True,
):
    """Build ViT-B32. All arguments passed to build_model."""
    if pretrained_top:
        assert classes == 1000, "Can only use pretrained_top if classes = 1000."
        assert include_top, "Can only use pretrained_top with include_top."
        assert pretrained, "Can only use pretrained_top with pretrained."
    model = build_model(
        **CONFIG_B,
        patch_size=32,
        image_size=image_size,
        classes=classes,
        activation=activation,
        include_top=include_top,
        representation_size=768 if pretrained and not pretrained_top else None,
    )
    if pretrained:
        load_pretrained(key="B32", model=model, pretrained_top=pretrained_top)
    return model


def vit_l16(
    image_size: int = 384,
    classes=1000,
    activation="linear",
    include_top=True,
    pretrained=True,
    pretrained_top=True,
):
    """Build ViT-L16. All arguments passed to build_model."""
    if pretrained_top:
        assert classes == 1000, "Can only use pretrained_top if classes = 1000."
        assert include_top, "Can only use pretrained_top with include_top."
        assert pretrained, "Can only use pretrained_top with pretrained."
    model = build_model(
        **CONFIG_L,
        patch_size=16,
        image_size=image_size,
        classes=classes,
        activation=activation,
        include_top=include_top,
        representation_size=None,
    )
    if pretrained:
        load_pretrained(key="L16", model=model, pretrained_top=pretrained_top)
    return model


def vit_l32(
    image_size: int = 384,
    classes=1000,
    activation="linear",
    include_top=True,
    pretrained=True,
    pretrained_top=True,
):
    """Build ViT-L32. All arguments passed to build_model."""
    if pretrained_top:
        assert classes == 1000, "Can only use pretrained_top if classes = 1000."
        assert include_top, "Can only use pretrained_top with include_top."
        assert pretrained, "Can only use pretrained_top with pretrained."
    model = build_model(
        **CONFIG_L,
        patch_size=32,
        image_size=image_size,
        classes=classes,
        activation=activation,
        include_top=include_top,
        representation_size=1024 if pretrained and not pretrained_top else None,
    )
    if pretrained:
        load_pretrained(key="L32", model=model, pretrained_top=pretrained_top)
    return model
